packages/terra-worklist-data-grid/src/WorklistDataGrid.jsx

```diff
@@ -320,31 +334,95 @@ function WorklistDataGrid(props) {
+      // The range extends upward from the anchor row
+      if (newEndRowIndex > selectionData.previousSelectionEndRow) {
+        // The range was moved down towards the anchor so rows that no longer qualify for the range need to be unselected.
+        const rowsToUnselect = rows.slice(selectionData.previousSelectionEndRow - 1, newEndRowIndex);
+        rowIdsToUnselect = rowsToUnselect.map(r => ({ rowId: r.id, isSelected: false }));
+      }
+    } else if (selectionData.anchorRow < selectionData.previousSelectionEndRow) {
+      // The range extends downward from the anchor row
+      if (newEndRowIndex < selectionData.previousSelectionEndRow) {
+        // The range was moved up towards the anchor so rows that no longer qualify for the range need to be unselected       // New endRangeIndex becomes ordered End
+        const rowsToUnselect = rows.slice(newEndRowIndex - 1, selectionData.previousSelectionEndRow);
+        rowIdsToUnselect = rowsToUnselect.map(r => ({ rowId: r.id, isSelected: false }));
+      }
+    }
+
+    let ariaMessage = intl.formatMessage({
+      id: hasSelectableRows
+        ? 'Terra.worklist-data-grid.row-selection-multiple-rows-selected'
+        : 'Terra.worklist-data-grid.row-selection-mode-enabled',
+    });
+
+    ariaMessage += intl.formatMessage({
+      id: rowIdsToSelect.length === 1
+        ? 'Terra.worklist-data-grid.row-selection-template'
+        : 'Terra.worklist-data-grid.row-selection-selected-rows-range',
+    }, { row: selectionData.anchorRow, endRow: newEndRowIndex });
+
+    setAriaLiveMessage(ariaMessage);
+    onRowSelect(rowIdsToSelect.concat(rowIdsToUnselect));
   };
 
   const selectRows = (selectAllRows, rowId, rowIndex) => {
     let rowLabel;
+    let isSelectAction = true;
     let msgId = 'Terra.worklist-data-grid.all-rows-selected';
+    // Reset last selected row when all rows are selected.
+    rowSelection.current.lastIndividuallySelectedRow = null;
     if (!selectAllRows) {
-      const isSelectAction = !rows[rowIndex - 1].isSelected; // Determine if this is select or unselected.
-      msgId = isSelectAction ? 'Terra.worklist-data-grid.row-selection-template' : 'Terra.worklist-data-grid.row-selection-cleared-template';
+      isSelectAction = !rows[rowIndex - 1].isSelected; // Determine if this is select or unselected.
+      // Remember the last selected row
+      rowSelection.current.lastIndividuallySelectedRow = { anchorRow: rowIndex };
+      if (isSelectAction) {
+        msgId = 'Terra.worklist-data-grid.row-selection-template';
+      } else {
+        msgId = 'Terra.worklist-data-grid.row-selection-cleared-template';
+      }
       rowLabel = rows[rowIndex - 1].ariaLabel || (rowIndex + 1);
     }
     setAriaLiveMessage(intl.formatMessage({ id: msgId }, { row: rowLabel }));
     if (selectAllRows && onRowSelectAll) {
       onRowSelectAll();
     } else if (onRowSelect) {
-      onRowSelect(rowId);
+      onRowSelect([{ rowId, isSelected: isSelectAction }]);
```

Comment:
For a single selection, I would expect that the consumer would be managing the selection state for the component.

Reply:
Updated